parsers/ELEXON.py

```diff
@@ -0,0 +1,202 @@
+        hours=span[0]).format('YYYY-MM-DD')
+    params['ToDate'] = target_datetime.replace(
+        hours=span[1]).format('YYYY-MM-DD')
+
+    if 'ELEXON_TOKEN' not in os.environ:
+        raise Exception('No ELEXON_TOKEN found! Please add it to secrets.env!')
+    params['APIKey'] = os.environ['ELEXON_TOKEN']
+    return session.get(ELEXON_ENDPOINT.format(report), params=params)
+
+
+def query_exchange(session, target_datetime=None):
+    params = {
+        'ServiceType': 'csv'
+    }
+    response = query_ELEXON('INTERFUELHH', session, params, target_datetime)
+    response = response.content.decode()
+    return response
+
+
+def query_production(session, target_datetime=None):
+    params = {
+        'ServiceType': 'csv'
+    }
+    response = query_ELEXON('FUELINST', session, params, target_datetime)
+    response = response.content.decode()
+    return response
+
+
+def parse_exchange(zone_key1, zone_key2, csv_text, target_datetime=None,
+                   logger=logging.getLogger(__name__)):
+    if not csv_text:
+        return None
+
+    sorted_zone_keys = sorted([zone_key1, zone_key2])
+
+    exchange = '->'.join(sorted_zone_keys)
+    data_points = list()
+    lines = csv_text.split('\n')
+
+    # check field count in report is as expected
+    field_count = len(lines[1].split(','))
+    if field_count != EXPECTED_FIELDS['INTERFUELHH']:
+        logger.warning(
+            'Expected {} fields in INTERFUELHH report, got {}'.format(
+                EXPECTED_FIELDS['INTERFUELHH'], field_count
+            )
+        )
+
+    for line in lines[1:-1]:
+        fields = line.split(',')
+
+        # create the datetime
+        datetime = datetime_from_date_sp(fields[1], fields[2])
+
+        data = {
+            'sortedZoneKeys': exchange,
+            'datetime': datetime,
+            'source': 'bmreports.com'
+        }
+
+        # positive value implies import to GB
+        multiplier = -1 if 'GB' in sorted_zone_keys[0] else 1
+        data['netFlow'] = float(fields[EXCHANGES[exchange]]) * multiplier
+        data_points.append(data)
+
+    if not target_datetime:
+        return data_points[-1]
+    else:
+        return data_points
+
+
+def parse_production(csv_text, target_datetime=None,
+                     logger=logging.getLogger(__name__)):
+    if not csv_text:
+        return None
+
+    data_points = list()
+    lines = csv_text.split('\n')
+
+    # check field count in report is as expected
+    field_count = len(lines[1].split(','))
+    if field_count != EXPECTED_FIELDS['FUELINST']:
+        logger.warning(
+            'Expected {} fields in FUELINST report, got {}'.format(
+                EXPECTED_FIELDS['FUELINST'], field_count
+            )
+        )
+
+    for line in lines[1:-1]:
+        fields = line.split(',')
+
+        # create the datetime
+        date_obj = dt.datetime.strptime(fields[3], '%Y%m%d%H%M%S')
+        datetime = arrow.get(date_obj).replace(minutes=-5).to('Europe/London')
```

Comment:
I'm still a bit puzzled by this. Doesn't the Elexon datetime represent the (start) date of the datapoint?

Reply:
Normally, yes. Note fields[3] is the publish time.

![image](https://user-images.githubusercontent.com/6525086/39698503-fdd9123e-51ec-11e8-9db6-f74de92e0ea6.png)
Settlement period 21 starts at 09:00GMT. For this settlement period, the first publication datetime is 09:05GMT and the last publication datetime is 09:30GMT. So I presume each publication datetime is referencing a datapoint with start date five minutes before that as the 5 minute observations for period 21 should be [09:00, 09:05, 09:10, 09:15, 09:20, 09:25] GMT. 
That's why I'm subtracting five minutes. Is that clear/correct?

Reply:
Makes sense! Can you add that comment in the code to make it obvious for future maintainers?
Thanks!